python/ray/serve/_private/metrics_utils.py

```diff
@@ -1,138 +1,103 @@
 from collections import defaultdict
 from dataclasses import dataclass, field
-from typing import Any, Callable, DefaultDict, Dict, List, Optional, Union
+from typing import Callable, DefaultDict, Dict, List, Optional
 
-import ray
-from ray.serve._private.constants import SERVE_LOGGER_NAME
+from ray.serve._private.constants import (
+    METRICS_PUSHER_GRACEFUL_SHUTDOWN_TIMEOUT_S,
+    SERVE_LOGGER_NAME,
+)
 
 logger = logging.getLogger(SERVE_LOGGER_NAME)
 
 
-class _MetricTask:
-    def __init__(self, task_func, interval_s, callback_func):
-        """
-        Args:
-            task_func: a callable that MetricsPusher will try to call in
-                each loop. It should take no arguments, and return a
-                kwargs dictionary to be used to call the callback_func.
-            interval_s: the interval of each task_func is supposed to be called.
-            callback_func: callback function is called when task_func is done, and
-                the result of task_func is passed to callback_func as the first
-                argument, and the timestamp of the call is passed as the second
-                argument.
-        """
-        self.task_func: Callable[[], Dict[str]] = task_func
-        self.interval_s: float = interval_s
-        self.callback_func: Callable[[Any, float]] = callback_func
-        self.last_ref: Optional[ray.ObjectRef] = None
-        self.last_call_succeeded_time: Optional[float] = time.time()
+@dataclass
+class _MetricsTask:
+    task_func: Callable
+    interval_s: float
 
 
 class MetricsPusher:
-    """
-    Metrics pusher is a background thread that run the registered tasks in a loop.
-    """
+    """Periodically runs registered asyncio tasks."""
 
-    # If no tasks are registered, sleep for 1s between iterations (until one is
-    # registered).
-    NO_TASKS_REGISTERED_INTERVAL_S = 1
+    def __init__(
+        self,
+        event_loop: asyncio.AbstractEventLoop,
+        async_sleep: Optional[Callable[[int], None]] = None,
+    ):
+        self._event_loop = event_loop
+        self._async_sleep = async_sleep or asyncio.sleep
+        self._tasks: Dict[str, _MetricsTask] = dict()
+        self._async_tasks: Dict[str, asyncio.Task] = dict()
+        self._stop_event = asyncio.Event(loop=event_loop)
 
-    def __init__(self):
-        self.tasks: Dict[str, _MetricTask] = dict()
-        self.pusher_thread: Union[threading.Thread, None] = None
-        self.stop_event = threading.Event()
+    def start(self):
+        self._stop_event.clear()
+
+    async def metrics_task(self, name: str):
+        """Periodically runs `task_func` every `interval_s` until `stop_event` is set.
+
+        If `task_func` raises an error, an exception will be logged.
+        """
+
+        wait_for_stop_event = asyncio.create_task(self._stop_event.wait())
+        while True:
+            if wait_for_stop_event.done():
+                return
+
+            try:
+                self._tasks[name].task_func()
```

Comment:
We need to check here that the task is still present, otherwise we're exposed to subtle RC with `stop_tasks` method, where we clear the set of tasks and then try to access the task in here

Reply:
Good point. In practice this should be very rare because we don't give up the asyncio loop after the `asyncio.wait` below, but we could have both of the timer and stop event tasks return at ~the same time

It'll be handled by the `except` block but the error log will be confusing. @zcin can you make a small PR handling the case where the dict is cleared?